File: ruby/question.rb
# implementation of the question function in Ruby
def question prompt, valid
  loop do
    puts prompt
    print "(#{valid.join ", "})" if valid.size != 0
    print ": "
    input = gets.chomp

    if valid.empty? || valid.include?(input)
      return input
    else
      puts %("#{input}" is not a valid answer)
    end
  end
end

question "foo", %w(bar baz)

File: elixir/question.ex
# question implementation in elixir
defmodule Question do
  def question(prompt), do: IO.gets("#{prompt}: ")

  def question(prompt, valid) do
    input = String.trim(IO.gets("#{prompt}\n(#{Enum.join(valid, ", ")}): "))

    if input in valid do
      input
    else
      IO.puts("\"#{input}\" is not a valid answer")
      question(prompt, valid)
    end
  end
end

Question.question("foo", ["bar", "baz"])
